### PR TITLE
Logic implemented

### PR DESCRIPTION
**Pull request by spioupiou**
FYI - I had to create a new route for the join method in the instance controller (this join method takes a pin submitted by a user as params)

I used 100_000 + instance.id as a pin (I am not sure but I don't think I can start from 0 because the pin column is an integer, if I use 000001 for example, active:record will most likely only record 1 as the pin)

When the lobby is not found, I re-render the page with a flash alert (bottom right corner). 
Of course we can implement AJAX later on but I don't think my front end can get any better than this lol

Would appreciate any feedback! 

https://user-images.githubusercontent.com/68413600/153752747-15548ede-4cd1-4f99-883d-fdb0240cdf99.mov

### CODE REVIEW

File: app/controllers/instances_controller.rb [hunk 1]
```diff
@@ -21,6 +25,15 @@ def create # Create the game instance
     end
   end
 
+  def join
+    @instance = Instance.find_by_pin(params[:pin])
+    if @instance.blank?
+      redirect_to root_path, notice: "Lobby not found"
+    else
+      redirect_to instance_path(@instance)
+    end
```
Comment (carlnoval) [inline app/controllers/instances_controller.rb:34]: If an instance is found then we should also create a player, player must be create with the matching instance_id.

File: app/views/games/index.html.erb [hunk 2]
```diff
@@ -15,6 +15,10 @@
           <% end %>
 
           <div class="btn btn-primary">Join a Lobby</div>
+          <%= form_with url: join_instances_path, method: :post do |f| %>
+            <%= f.text_field :pin, value: params[:pin] %>
+            <%= f.submit "Search"%>
```
Comment (carlnoval) [inline app/views/games/index.html.erb:20]: Played with it and it works however this kinda gives us a third button now 😅 . I am not sure how we can turn it to just two buttons ideally the `create` and `join` buttons only. @hiroya210 any suggestions, hoping that you got something that would also make it easy for the front-end to manipulate.

File: config/routes.rb [hunk 3]
```diff
@@ -3,5 +3,8 @@
   root to: 'games#index' # our landing page with selection of games
   resources :instances, only: [:create, :show, :update] do
     resources :player_inputs, only: [:create, :index]
+    collection do
+      post :join # to join the instance via pin
+    end
   end
 end
```
Comment (carlnoval) [inline config/routes.rb:10]: 🤔 Goal of the route in the end is to create a player. I think we may as well create `resources :player, only: [:create]` below like 9. In the future this will also be handy to have a destroy player if ever we decide to allow the player to leave a game instance before the game starts.

File: app/controllers/instances_controller.rb [hunk 4]
```diff
@@ -8,6 +8,10 @@ def create # Create the game instance
         status: "pending"
       )
 
+      # Simplistic pin number
+      @instance.pin = 100_000 + @instance.id
+      @instance.save
```
Comment (carlnoval) [inline app/controllers/instances_controller.rb:13]: I think this is good and simple for now. Later on maybe we can create our own pin mechanism or use something like this: https://github.com/spannerinworks/easy-pin/.

**Comment by spioupiou** [comment]
I've made a few modifications, thanks again @carlnoval for your advice!
- I changed the route using `resources :players, only: [:create]`
- In the index: still ugly but I got rid of the duplicate button, now it's only "Create" and the form to join a room
- Added a players_controller and moved my "join" method there, renamed it as "create"
- When user enters the correct pin, he automatically becomes a player before being redirected to the instance show page
- In the instance show page I added the list of players and the pin code for clarity
- Simplified instances_controller#create by using `redirect_to new_user_session_path unless user_signed_in?` instead of nested if/else statements

Screen Shot 2022-02-14 at 9 49 06
Screen Shot 2022-02-14 at 10 35 26

File: app/controllers/players_controller.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+class PlayersController < ApplicationController
+  def create
+    redirect_to new_user_session_path unless user_signed_in?
```
Comment (spioupiou) [inline app/controllers/players_controller.rb:3]: This statement doesn't work... which makes no sense because it works perfectly fine in the other controller (instances#create) I have the feeling this is an issue with form_for, but I have no idea how to change it. I would suggest that we proceed as the code works when we are logged in - and I'll grab a teacher on Tuesday to ask for help

File: app/controllers/players_controller.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+    @instance = Instance.find_by_pin(params[:pin])
+    if @instance.blank?
+      redirect_to root_path, notice: "Lobby not found"
+    else # Need some logic to not create dupplicate users (will implement later)
```
Comment (spioupiou) [inline app/controllers/players_controller.rb:8]: While testing I realize that we have no logic to prevent duplicate players - I think we can add validations in model and some code here, but that will be for another PR!

Comment (carlnoval) [reply app/controllers/players_controller.rb:8]: Sounds good to me, created this Trello so we don't forget about this: https://trello.com/c/ZsoGGTzt/23-dbi-prevent-the-same-player-to-be-created-again

File: app/controllers/players_controller.rb (hunk 5 above)
Comment (carlnoval) [reply app/controllers/players_controller.rb:3]: Ok, also created a Trello for this so we don't lose track: https://trello.com/c/N0Pssf25/24-dbi-after-clicking-create-redirect-to-login-page-if-user-is-not-logged-in

**Comment by carlnoval** [comment]
This now looks good to me, merging.
